sycl/doc/extensions/proposed/sycl_ext_oneapi_prefetch.asciidoc

```diff
@@ -0,0 +1,454 @@
+
+enum class cache_level {
+  L1,
+  L2,
+  L3,
+  L4,
+};
+
+struct nontemporal;
+
+struct prefetch_hint_key {
+  template <cache_level Level, typename Hint>
+  using value_t = property_value<prefetch_hint_key,
+                                 std::integral_constant<cache_level, Level>,
+                                 Hint>
+};
+
+template <cache_level Level, typename Hint>
+inline constexpr prefetch_hint_key::value_t<Level, Hint> prefetch_hint;
+
+inline constexpr prefetch_hint_key::value_t<cache_level::L1, void> prefetch_hint_L1;
+inline constexpr prefetch_hint_key::value_t<cache_level::L2, void> prefetch_hint_L2;
+inline constexpr prefetch_hint_key::value_t<cache_level::L3, void> prefetch_hint_L3;
+inline constexpr prefetch_hint_key::value_t<cache_level::L4, void> prefetch_hint_L4;
+
+inline constexpr prefetch_hint_key::value_t<cache_level::L1, nontemporal> prefetch_hint_L1_NT;
+inline constexpr prefetch_hint_key::value_t<cache_level::L2, nontemporal> prefetch_hint_L2_NT;
+inline constexpr prefetch_hint_key::value_t<cache_level::L3, nontemporal> prefetch_hint_L3_NT;
+inline constexpr prefetch_hint_key::value_t<cache_level::L4, nontemporal> prefetch_hint_L4_NT;
+
+} // namespace sycl::ext::oneapi::experimental
+----
+
+If the `nontemporal` type is used as the value of the `Hint` template
+parameter, this acts as a hint that the data being prefetched will not be
+reused.
+
+[NOTE]
+====
+The implementation of `nontemporal` is device-specific, owing to differences in
+cache hierarchies and replacement policies. On some devices non-temporal data
+may be evicted earlier than temporal data, while on other devices non-temporal
+data may not be stored in certain levels of cache at all.
+====
+
+
+=== Work-item prefetches
+
+The functions in this section allow individual work-items to prefetch data.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename Properties = ext::oneapi::experimental::properties<>>
+void prefetch(void* ptr, Properties properties = {});
+
+template <typename Properties = ext::oneapi::experimental::properties<>>
+void prefetch(void* ptr, size_t bytes, Properties properties = {});
+
+template <typename T, typename Properties = ext::oneapi::experimental::properties<>>
+void prefetch(T* ptr, Properties properties = {});
+
+template <typename T, typename Properties = ext::oneapi::experimental::properties<>>
+void prefetch(T* ptr, size_t count, Properties properties = {});
+
+// Only available if Accessor::dimensions == 1 and Accessor::access_mode != write
+template <typename Accessor, typename Properties = ext::oneapi::experimental::properties<>>
+void prefetch(Accessor* acc, size_t offset, Properties properties = {});
+
+// Only available if Accessor::dimensions == 1 and Accessor::access_mode != write
+template <typename Accessor, typename Properties = ext::oneapi::experimental::properties<>>
+void prefetch(Accessor* acc, size_t offset, size_t count, Properties properties = {});
```

Comment:
I think there should be some clarification on which types are allowed for `Accessor`.  For example:

* `accessor` with `target::device`.  (This excludes `target::host_task` and all of the deprecated targets.)
* `local_accessor`.

I assume you want to exclude the image accessors?  You certainly want to exclude the host accessors.

It seems like this would be more clear if there were different overloads for `accessor` and `local_accessor`, rather than a long English description on the constraints for `Accessor`.

Reply:
I think `accessor` with `target::device` is the only one we want to support here.  I can't immediately think of any devices or backends where developers would need or want to prefetch from local memory, and the SYCL spec only talks about a "global memory cache".

Reply:
I've switched things to use `accessor` in 62faa8bd1fc5ee18272b45bec36f6850180e9cf3 and directly limited it to `target::device`, which simplifies the constraints.  I'm going to make the dimensionality change you suggested in a follow-up commit, and that should simplify things further.